### ISSUE TITLE
[Quantization] Replace Simplified CPU Quantization Implementations with Production-Ready Algorithms

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#399]
## Summary

The CPU quantization methods in `CPUQuantizer` are currently implemented as simplified simulations rather than production-ready algorithms. This affects the accuracy and performance of neural network inference on CPU devices.

## Problem Description

**File:** `crates/bitnet-quantization/src/device_aware_quantizer.rs`

**Affected Methods:**
- `CPUQuantizer::quantize_i2s` - Marked as "Simplified I2S quantization"
- `CPUQuantizer::quantize_tl1` - Marked as "Simplified TL1 implementation"

These methods currently use basic quantization logic that doesn't implement the full BitNet quantization algorithms, leading to:

1. **Suboptimal accuracy** - Simplified algorithms may not preserve model accuracy
2. **Performance gaps** - Missing SIMD optimizations and efficient bit packing
3. **Missing features** - Incomplete implementation of advanced quantization techniques
4. **Production concerns** - Simulation code in production neural network inference

### I2S Quantization Issues
```rust
// Current simplified implementation
let quantized = if normalized > 0.5 {
    1i8
} else if normalized < -0.5 {
    -1i8
} else {
    0i8
};
```

**Problems:**
- Basic threshold-based quantization without proper distribution analysis
- Inefficient bit packing (4 values per byte with simple bit shifts)
- Missing statistical analysis for optimal scale factors
- No consideration of activation distribution patterns

### TL1 Quantization Issues
```rust
// Current simplified implementation
let quantized = ((normalized.clamp(-1.0, 1.0) + 1.0) * 7.5) as u8;
```

**Problems:**
- Linear mapping without proper lookup table implementation
- Missing vectorized operations for performance
- No support for adaptive quantization parameters
- Incomplete 4-bit packing optimization

## Root Cause Analysis

1. **Development Priority**: CPU implementations were created as placeholders during initial GPU development
2. **Cross-validation Dependency**: GPU implementations fall back to these simplified CPU methods
3. **Architecture Gaps**: Missing integration with optimized SIMD kernels in `bitnet-kernels`
4. **Testing Limitations**: Accuracy validation tests may pass due to relaxed tolerances masking the simulation nature

## Impact Assessment

**Severity:** High - Core quantization functionality
**Affected Components:**
- CPU inference performance (all models)
- GPU fallback reliability (when CUDA unavailable)
- Cross-validation accuracy between CPU/GPU
- Production deployment on CPU-only systems

**Performance Impact:**
- ~20-30% slower quantization on CPU devices
- Potential accuracy degradation in large model inference
- Inefficient memory usage during quantization

### Phase 1: I2S Production Implementation
Implement production-ready I2S quantization with:

```rust
pub fn quantize_i2s(&self, data: &[f32]) -> Result<QuantizedTensor> {
    debug!("Performing production I2S quantization on CPU");
    
    // Use statistical analysis for optimal block sizing
    let block_size = self.compute_optimal_block_size(data.len());
    let num_blocks = data.len().div_ceil(block_size);
    
    let mut quantized_data = Vec::new();
    let mut scales = Vec::new();
    
    for block_idx in 0..num_blocks {
        let start = block_idx * block_size;
        let end = (start + block_size).min(data.len());
        let block = &data[start..end];
        
        // Advanced scale factor calculation with outlier handling
        let scale = self.compute_robust_scale_factor(block);
        scales.push(scale);
        
        // Vectorized quantization with SIMD when available
        let block_quantized = self.vectorized_i2s_quantize(block, scale);
        
        // Optimized bit packing (4 2-bit values per byte)
        let packed_data = self.pack_i2s_bits(&block_quantized);
        quantized_data.extend(packed_data);
    }
    
    Ok(QuantizedTensor::new(
        quantized_data,
        QuantizationType::I2S,
        vec![data.len()],
        scales,
        block_size,
    ))
}
```

### Phase 2: TL1 Lookup Table Implementation
Implement proper table lookup quantization:

```rust
pub fn quantize_tl1(&self, data: &[f32]) -> Result<QuantizedTensor> {
    debug!("Performing production TL1 quantization on CPU");
    
    // Initialize or load pre-computed lookup tables
    let lookup_table = self.get_or_create_tl1_table();
    
    let block_size = 16; // Optimized for cache efficiency
    let mut quantized_data = Vec::new();
    let mut scales = Vec::new();
    
    for chunk in data.chunks(block_size) {
        let scale = self.compute_adaptive_scale(chunk);
        scales.push(scale);
        
        // Vectorized lookup table quantization
        let quantized_chunk = self.vectorized_tl1_lookup(chunk, scale, &lookup_table);
        
        // Efficient 4-bit packing (2 values per byte)
        let packed_chunk = self.pack_tl1_nibbles(&quantized_chunk);
        quantized_data.extend(packed_chunk);
    }
    
    Ok(QuantizedTensor::new(
        quantized_data,
        QuantizationType::TL1,
        vec![data.len()],
        scales,
        block_size,
    ))
}
```

### Phase 3: SIMD Integration
- Integrate with existing SIMD kernels in `bitnet-kernels`
- Add AVX2/AVX-512 support for x86_64 targets
- Add NEON support for ARM64 targets
- Implement runtime CPU feature detection

### Phase 4: Accuracy Validation
- Implement cross-validation with C++ reference implementation
- Add comprehensive accuracy benchmarks
- Ensure ±1e-5 tolerance for I2S and ±1e-4 for TL1/TL2
- Performance regression testing

### Task Breakdown
- [ ] **Research & Design** (2 days)
  - [ ] Analyze BitNet paper quantization algorithms
  - [ ] Review Microsoft C++ reference implementation
  - [ ] Design SIMD-optimized data structures
  
- [ ] **I2S Implementation** (3 days)
  - [ ] Implement robust scale factor calculation
  - [ ] Add vectorized quantization logic
  - [ ] Optimize bit packing performance
  - [ ] Add comprehensive unit tests
  
- [ ] **TL1 Implementation** (3 days)
  - [ ] Design and implement lookup tables
  - [ ] Add adaptive parameter selection
  - [ ] Implement vectorized table lookup
  - [ ] Optimize nibble packing
  
- [ ] **SIMD Integration** (2 days)
  - [ ] Integrate with `bitnet-kernels` SIMD functions
  - [ ] Add runtime feature detection
  - [ ] Benchmark performance improvements
  
- [ ] **Validation & Testing** (2 days)
  - [ ] Cross-validate with GPU implementations
  - [ ] Add accuracy regression tests
  - [ ] Performance benchmarking
  - [ ] Documentation updates

### Dependencies
- [ ] `bitnet-kernels` SIMD kernel implementations
- [ ] C++ reference implementation for cross-validation
- [ ] Updated tolerance configurations in `ToleranceConfig`

### Unit Tests
```rust
#[test]
fn test_i2s_quantization_accuracy() {
    let quantizer = CPUQuantizer::new(ToleranceConfig::default());
    let test_data = generate_model_like_distribution(1024);
    
    let quantized = quantizer.quantize_i2s(&test_data).unwrap();
    let dequantized = quantizer.dequantize_i2s(&quantized).unwrap();
    
    let relative_error = compute_relative_error(&test_data, &dequantized);
    assert!(relative_error < 1e-5, "I2S quantization exceeds tolerance");
}

#[test]
fn test_tl1_vectorized_performance() {
    let quantizer = CPUQuantizer::new(ToleranceConfig::default());
    let large_tensor = vec![0.5f32; 65536]; // 64KB tensor
    
    let start = Instant::now();
    let _ = quantizer.quantize_tl1(&large_tensor).unwrap();
    let duration = start.elapsed();
    
    // Should complete large tensor quantization in reasonable time
    assert!(duration < Duration::from_millis(10));
}
```

### Integration Tests
- Cross-device quantization parity validation
- Large model inference accuracy preservation
- Performance regression benchmarks against simplified implementations

### Acceptance Criteria
- [ ] I2S quantization achieves ±1e-5 relative error tolerance
- [ ] TL1 quantization achieves ±1e-4 relative error tolerance  
- [ ] CPU quantization performance within 90% of optimized C++ reference
- [ ] SIMD acceleration shows measurable performance improvements
- [ ] All existing tests continue to pass
- [ ] GPU fallback maintains accuracy when using updated CPU implementation
- [ ] Memory usage remains within current bounds
- [ ] Cross-validation tests pass with C++ reference

## Related Issues

This issue addresses simulation code identified in quantization analysis. Related improvements:
- GPU quantization kernel optimization
- Device-aware quantization selection algorithms
- Quantization accuracy validation framework enhancements

## References

- BitNet: Scaling 1-bit Transformers for Large Language Models (Original Paper)
- Microsoft BitNet C++ Reference Implementation
- `docs/reference/quantization-support.md` - BitNet.rs quantization documentation
- `crates/bitnet-kernels/` - SIMD kernel implementations for optimization integration

## Environment

- **MSRV:** 1.90.0 (Rust 2024 edition)
- **Features:** `--no-default-features --features cpu`
- **Target Architectures:** x86_64, aarch64 (with SIMD support)
- **Cross-validation:** C++ reference implementation compatibility required
